### DISCUSSION TITLE
working with shaders in py5

### DISCUSSION THREAD

**Discussion by Maltiec** [py5coding/py5generator#300] [Q&A] [answered]
Has anyone worked with shaders in py5? is there any special rules for shaders which will be used in apply_filter()? I've tried some examples with calling py5.shader() and using set method of Shader class - and it was a success, but apply_filter() works in a very strange way (black window most of the time).

**Comment by hx2A** [comment]
Yes, I have used shaders extensively, and have spent much of today with that. `apply_filter()` should also work. Not sure why you are getting a black window. Can you provide some examples? What shader are you passing to `apply_filter()`?

**Reply by hx2A** [reply, marked as answer]
I looked at some code I wrote that uses `apply_filter()` and I recall the shader needs to work a bit differently than regular shaders. Normally when you create a vert and frag shader it is for drawing triangles and the shader has no information about pixels it is being drawn on top of. `apply_filter()` is different in that you have access to the all of the pixels in the window.

The following fragment shader could be used with `apply_filter()`.

```c
// filename: data/red.glsl

// processing supplied variables
uniform sampler2D texture;
varying vec4 vertColor;
varying vec4 vertTexCoord;

void main() {
  // get original color for this pixel
  vec4 originalColor = texture2D(texture, vertTexCoord.st) * vertColor;

  // remove green and blue, leaving red
  vec4 color = vec4(originalColor.r, 0.0, 0.0, 1.0);

  // gl_FragColor needs to be set to something
  gl_FragColor = color;
}
```

Test code:

```python
def setup():
    py5.size(400, 400, py5.P2D)
    py5.rect(10, 10, 300, 300)
    shader = py5.load_shader('red.glsl')
    py5.apply_filter(shader)
```

Shaders take some time and effort to understand, and Processing Shaders add some complexity because of the Processing framework. Once you get the hang of them, they are super useful.

**Reply by Maltiec** [reply]
Thanks a lot! I've tried some examples from https://webgl-shaders.com/  And those shaders really differ from processing-specific ones.

**Reply by hx2A** [reply]
You are welcome.

Those shaders might work with Processing but will take some effort and knowledge of what Processing Shaders need & want to integrate them into Processing's framework.
